### ISSUE TITLE
[Feature Request] SuperLoss

### ISSUE THREAD

**Issue by kayuksel** [pytorch/pytorch#49851]
## 🚀 Feature
SuperLoss is a simple and generic method that can be applied to a variety of losses and tasks without any changing the learning procedure. It consists in appending a novel loss function on top of any existing task loss. Its main effect is to automatically down-weight the contribution of samples with a large loss,i.e. hard samples, effectively mimicking the curriculum learning. SuperLoss prevents the memorization of noisy samples, making it possible to train from noisy data even with non-robust loss functions. Experimental results on image classification, regression, object detection and image retrieval demonstrated consistent gain.

## Motivation
https://proceedings.neurips.cc//paper/2020/file/2cfa8f9e50e0f510ede9d12338a5f564-Paper.pdf
https://proceedings.neurips.cc/paper/2020/file/2cfa8f9e50e0f510ede9d12338a5f564-Supplemental.pdf

**Comment by mruberry** [comment]
Hey @kayuksel, thanks for the suggestion. Given how new SuperLoss is it should probably be implemented in a PyTorch-compatible library.

**Comment by kayuksel** [comment]
@mruberry Thank you very much for moving forward.

Authors said that they have used "a precomputed piece-wise approximation of the (Lambert W) function, which can be easily implemented on GPU in PyTorch using the grid_sample() function."

I also found two implementations of Lambert W in PyTorch, I am sharing those in case that they could be helpful:

https://github.com/thibsej/unbalanced-ot-functionals/blob/13f2203b3993d973f929578085ea458c5c1a7a78/common/torch_lambertw.py

https://github.com/AminJun/BreakingCertifiableDefenses/blob/cc469fa48f7efba21f3584e233c4db0c9a4856c1/RandomizedSmoothing/projected_sinkhorn/lambertw.py

**Comment by mruberry** [comment]
@kayuksel That's great, but to be clear I'm suggesting that this be implemented in a PyTorch-compatible library at this time, not in PyTorch Core. There are a lot of libraries built on top of PyTorch like GPyTorch and Captum. New research like this typically incubates in them, and then a determination is made for whether it's fundamental enough to be in core or remain in a more domain-specific library. See the Developer FAQ which elaborates on this: https://github.com/pytorch/pytorch/wiki/Developer-FAQ#i-have-a-new-function-or-feature-id-like-to-add-to-pytorch-should-it-be-in-pytorch-core-or-a-library-like-torchvision.

**Comment by gmgeorg** [comment]
@kayuksel probably late to the party, but nonetheless worth mentioning that I just published a torch native implementation of the Lambert W function in the **torchlambertw** module:
https://github.com/gmgeorg/torchlambertw
